File: packages/a2a/src/card-mapper.ts
/**
 * PolyMesh capability ↔ A2A skill mapping (PM-V6-SPEC §A.5.1–§A.5.3).
 *
 * Outbound M2 only needs the forward direction (skill naming + fidelity
 * clause); the reverse helper is provided for M3 inbound advertisement import.
 */
import type { A2ASkill, PolyMeshCapability } from "./types.js";

export const POLYMESH_CAPABILITY_PREFIX = "org.polymesh.";

/** Fidelity clause prefix that MUST appear in every projected description. */
export const FIDELITY_CLAUSE_PREFIX = "PolyMesh capability id: ";

/**
 * §A.5.1 strip rule. Only the `org.polymesh.` prefix is stripped; every other
 * reverse-DNS id passes through unchanged.
 */
export function skillNameFromCapabilityName(name: string): string {
  if (name.startsWith(POLYMESH_CAPABILITY_PREFIX)) {
    return name.slice(POLYMESH_CAPABILITY_PREFIX.length);
  }
  return name;
}

/** Alias matching the spec pseudocode identifier. */
export const skill_name_from_capability_name = skillNameFromCapabilityName;

/** Build the fidelity clause the description MUST contain (§A.5.1). */
export function fidelityClause(capabilityName: string): string {
  return `${FIDELITY_CLAUSE_PREFIX}${capabilityName}`;
}

/** True when `description` carries the fidelity clause for `capabilityName`. */
export function hasFidelityClause(description: string, capabilityName: string): boolean {
  return description.includes(fidelityClause(capabilityName));
}

/**
 * Extract the full PolyMesh capability id from a skill description
 * (reverse lookup, §A.5.2). Returns `undefined` when the clause is absent.
 */
export function capabilityIdFromDescription(description: string): string | undefined {
  const index = description.indexOf(FIDELITY_CLAUSE_PREFIX);
  if (index < 0) return undefined;
  const rest = description.slice(index + FIDELITY_CLAUSE_PREFIX.length);
  const match = /^[A-Za-z0-9_.:-]+/.exec(rest.trimStart());
  return match ? match[0] : undefined;
}

/** Project a PolyMesh capability contract onto an A2A skill (§A.5.1). */
export function capabilityToA2ASkill(capability: PolyMeshCapability): A2ASkill {
  const name = skillNameFromCapabilityName(capability.name);
  const base = capability.description?.trim();
  const description = base
    ? `${base} (${fidelityClause(capability.name)})`
    : fidelityClause(capability.name);

  const tags: string[] = [];
  if (capability.idempotency) tags.push(`idempotency=${capability.idempotency}`);
  if (capability.side_effects) tags.push(`side_effects=${capability.side_effects}`);
  if (capability.version) tags.push(`version=${capability.version}`);

  const metadata: Record<string, unknown> = {};
  if (capability.idempotency) metadata.idempotency = capability.idempotency;
  if (capability.side_effects) metadata.side_effects = capability.side_effects;
  if (capability.approval) metadata.approval = capability.approval;
  if (capability.cancellation) metadata.cancellation = capability.cancellation;
  if (capability.timeout_ceiling_seconds !== undefined) {
    metadata.timeout_ceiling_seconds = capability.timeout_ceiling_seconds;
  }

  return {
    id: capability.name,
    name,
    description,
    tags,
    inputModes: ["application/json"],
    outputModes: ["application/json"],
    ...(capability.input_schema ? { inputSchema: capability.input_schema } : {}),
    ...(capability.result_schema ? { outputSchema: capability.result_schema } : {}),
    ...(Object.keys(metadata).length > 0 ? { metadata } : {}),
  };
}

const REVERSE_DNS_PREFIX = /^(org|com|io|net|dev|custom)\./;

/**
 * Reverse mapping (§A.5.2): resolve the PolyMesh capability name advertised by
 * a remote A2A skill. Prefers the fidelity clause; otherwise re-attaches the
 * `org.polymesh.` prefix only for non reverse-DNS skill names.
 */
export function capabilityNameFromSkill(skill: {
  name: string;
  description?: string;
  origin?: string;
}): { name: string; origin: "polymesh" | "a2a_native" } {
  const declared = skill.description ? capabilityIdFromDescription(skill.description) : undefined;
  if (declared && (skill.origin === undefined || skill.origin === "polymesh")) {
    return { name: declared, origin: "polymesh" };
  }
  if (!REVERSE_DNS_PREFIX.test(skill.name)) {
    return { name: `${POLYMESH_CAPABILITY_PREFIX}${skill.name}`, origin: "a2a_native" };
  }
  return { name: skill.name, origin: "a2a_native" };
}

/** Capabilities that MUST NOT be published inbound without operator opt-in (§A.5.3). */
export const INBOUND_PUBLISH_DENYLIST: readonly string[] = Object.freeze([
  "org.polymesh.shell.exec",
  "org.polymesh.file.write",
]);

/** §A.5.3 publish gate. Inbound-only, retained here for M3 reuse. */
export function isPublishableSkill(capability: PolyMeshCapability): boolean {
  if (INBOUND_PUBLISH_DENYLIST.includes(capability.name)) return false;
  if (capability.approval === "always") return false;
  if (capability.side_effects === "approval") return false;
  return true;
}

/** Project a set of capabilities onto A2A skills, applying the §A.5.3 gate. */
export function mapCapabilitiesToSkills(
  capabilities: readonly PolyMeshCapability[],
  options?: { enforcePublishGate?: boolean },
): A2ASkill[] {
  const gate = options?.enforcePublishGate ?? true;
  return capabilities
    .filter((c) => (gate ? isPublishableSkill(c) : true))
    .map((c) => capabilityToA2ASkill(c));
}

export function skillDescriptionFromCapability(
  name: string,
  description?: string,
  version?: string,
): string {
  let desc = `${fidelityClause(name)}. `;
  if (description) desc += description;
  if (version) desc += (desc.endsWith(" ") ? "" : " ") + `version=${version}`;
  return desc.trim();
}

export function mapCapabilityToSkill(capability: PolyMeshCapability): A2ASkill {
  return capabilityToA2ASkill(capability);
}

export function mapCardToA2a(
  card: Record<string, unknown>,
  options?: { enforcePublishGate?: boolean },
): Record<string, unknown> {
  const caps = Array.isArray(card.capabilities)
    ? (card.capabilities as PolyMeshCapability[])
    : [];
  return {
    name: card.agent_id,
    version: card.card_version ?? "0.0.0",
    skills: mapCapabilitiesToSkills(caps, {
      enforcePublishGate: options?.enforcePublishGate ?? false,
    }),
  };
}

export function mapCardFromA2a(card: Record<string, unknown>): Record<string, unknown> {
  const skills = Array.isArray(card.skills) ? (card.skills as A2ASkill[]) : [];
  return {
    agent_id: card.name,
    card_version: card.version ?? "0.0.0-a2a",
    capabilities: skills.map((s) => ({
      name: capabilityNameFromSkill(s),
      description: s.description,
      dialect: "a2a" as const,
      a2a_url: typeof card.url === "string" ? card.url : undefined,
    })),
  };
}

export const map_card_to_a2a = mapCardToA2a;
export const map_card_from_a2a = mapCardFromA2a;

File: src/polymesh/a2a/card_mapper.py
"""PolyMesh AgentCard <-> A2A AgentCard mapping (§A.5).

M2 only needs the outbound direction (capability id -> skill name), so the card
level mappers are intentionally thin.  The skill naming rules are the normative
ones and are shared with the inbound work in M3.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from .types import A2AAgentCard, A2ASkill

POLYMESH_CAPABILITY_PREFIX = "org.polymesh."

#: Fidelity clause appended to skill descriptions so inbound can round-trip.
FIDELITY_CLAUSE = "PolyMesh capability id: "

#: Reverse-DNS-looking prefixes that must never be re-prefixed on the way back.
REVERSE_DNS_PREFIXES = ("org.", "com.", "io.", "net.", "dev.", "custom.")

#: Capabilities that MUST NOT be published inbound without operator opt-in (§A.5.3).
INBOUND_PUBLISH_DENYLIST: frozenset[str] = frozenset(
    {
        "org.polymesh.shell.exec",
        "org.polymesh.file.write",
    }
)


def is_publishable_skill(capability: Mapping[str, Any]) -> bool:
    """§A.5.3 publish gate for inbound AgentCard skills."""

    name = str(capability.get("name") or capability.get("capability") or "")
    if name in INBOUND_PUBLISH_DENYLIST:
        return False
    if capability.get("approval") == "always":
        return False
    if capability.get("side_effects") == "approval":
        return False
    return True


def map_capabilities_to_skills(
    capabilities: Sequence[Mapping[str, Any]],
    *,
    enforce_publish_gate: bool = True,
) -> list[A2ASkill]:
    skills: list[A2ASkill] = []
    for capability in capabilities:
        if enforce_publish_gate and not is_publishable_skill(capability):
            continue
        skills.append(map_capability_to_skill(capability))
    return skills


def skill_name_from_capability_name(name: str) -> str:
    """Strip the ``org.polymesh.`` prefix only (§A.5.1 normative algorithm)."""

    if not isinstance(name, str):
        raise TypeError("capability name must be a string")
    if name.startswith(POLYMESH_CAPABILITY_PREFIX):
        return name[len(POLYMESH_CAPABILITY_PREFIX) :]
    return name


def capability_name_from_skill_name(
    skill_name: str,
    *,
    description: str | None = None,
    origin: str | None = None,
) -> str:
    """Reverse the strip rule, preferring the fidelity clause (§A.5.1)."""

    if isinstance(description, str) and FIDELITY_CLAUSE in description:
        tail = description.split(FIDELITY_CLAUSE, 1)[1].strip()
        candidate = tail.split()[0].rstrip(".,;") if tail else ""
        if candidate:
            return candidate
    if origin not in (None, "polymesh"):
        return skill_name
    if skill_name.startswith(REVERSE_DNS_PREFIXES):
        return skill_name
    return POLYMESH_CAPABILITY_PREFIX + skill_name


def _skill_description(capability: Mapping[str, Any], capability_id: str) -> str:
    base = capability.get("description")
    text = str(base).strip() if isinstance(base, str) and base.strip() else capability_id
    if FIDELITY_CLAUSE in text:
        return text
    return f"{text} ({FIDELITY_CLAUSE}{capability_id})"


def map_capability_to_skill(capability: Mapping[str, Any]) -> A2ASkill:
    """Project one PolyMesh capability advertisement onto an A2A skill."""

    capability_id = str(capability.get("name") or capability.get("capability") or "")
    if not capability_id:
        raise ValueError("capability advertisement requires a name")
    skill: A2ASkill = {
        "name": skill_name_from_capability_name(capability_id),
        "description": _skill_description(capability, capability_id),
        "inputModes": ["application/json"],
        "outputModes": ["application/json"],
    }
    input_schema = capability.get("input_schema")
    if isinstance(input_schema, Mapping):
        skill["inputSchema"] = dict(input_schema)
    result_schema = capability.get("result_schema")
    if isinstance(result_schema, Mapping):
        skill["outputSchema"] = dict(result_schema)
    metadata: dict[str, Any] = {"polymesh_capability_id": capability_id}
    for field in ("version", "idempotency", "side_effects", "cancellation", "timeout_ceiling_seconds"):
        value = capability.get(field)
        if value is not None:
            metadata[field] = value
    skill["metadata"] = metadata
    return skill


def map_card_to_a2a(
    card: Mapping[str, Any],
    *,
    url: str | None = None,
    enforce_publish_gate: bool = False,
) -> A2AAgentCard:
    """Minimal PolyMesh AgentCard -> A2A AgentCard projection.

    Mesh topology, scope, and room membership are never copied (§A.16.3).
    Pass ``enforce_publish_gate=True`` for inbound AgentCard publishing (§A.5.3).
    """

    capabilities = card.get("capabilities")
    skills: list[A2ASkill] = []
    if isinstance(capabilities, Sequence) and not isinstance(capabilities, (str, bytes)):
        caps = [c for c in capabilities if isinstance(c, Mapping)]
        skills = map_capabilities_to_skills(caps, enforce_publish_gate=enforce_publish_gate)
    a2a_card: A2AAgentCard = {
        "name": str(card.get("display_name") or card.get("agent_id") or "polymesh-agent"),
        "description": str(card.get("description") or "PolyMesh agent exposed over the A2A dialect"),
        "version": str(card.get("version") or "1.0.0"),
        "capabilities": {"streaming": bool(card.get("sse_enabled", False))},
        "skills": skills,
    }
    resolved_url = url or card.get("a2a_url")
    if isinstance(resolved_url, str) and resolved_url:
        a2a_card["url"] = resolved_url
    return a2a_card


def map_card_from_a2a(card: Mapping[str, Any], *, a2a_url: str | None = None) -> dict[str, Any]:
    """Minimal A2A AgentCard -> PolyMesh advertisement projection (§A.5.2).

    Imported skills fail closed: ``sensitive`` idempotency, ``network`` side
    effects, and the ``a2a`` dialect tag with its ``a2a_url``.
    """

    resolved_url = a2a_url or card.get("url")
    skills = card.get("skills")
    capabilities: list[dict[str, Any]] = []
    if isinstance(skills, Sequence) and not isinstance(skills, (str, bytes)):
        for skill in skills:
            if not isinstance(skill, Mapping):
                continue
            skill_name = str(skill.get("name") or "")
            if not skill_name:
                continue
            description = skill.get("description")
            capability_id = capability_name_from_skill_name(
                skill_name,
                description=description if isinstance(description, str) else None,
                origin="polymesh" if isinstance(description, str) and FIDELITY_CLAUSE in description else "a2a_native",
            )
            entry: dict[str, Any] = {
                "name": capability_id,
                "description": str(description) if isinstance(description, str) else capability_id,
                "version": "1.0.0",
                "idempotency": "sensitive",
                "side_effects": "network",
                "dialect": "a2a",
                "input_schema": dict(skill["inputSchema"]) if isinstance(skill.get("inputSchema"), Mapping) else {"type": "object"},
                "result_schema": dict(skill["outputSchema"]) if isinstance(skill.get("outputSchema"), Mapping) else {"type": "object"},
            }
            if isinstance(resolved_url, str) and resolved_url:
                entry["a2a_url"] = resolved_url
            capabilities.append(entry)
    advertisement: dict[str, Any] = {
        "agent_id": str(card.get("name") or "a2a-remote"),
        "display_name": str(card.get("name") or "a2a-remote"),
        "capabilities": capabilities,
        "dialect": "a2a",
    }
    if isinstance(resolved_url, str) and resolved_url:
        advertisement["a2a_url"] = resolved_url
    return advertisement


__all__ = [
    "FIDELITY_CLAUSE",
    "INBOUND_PUBLISH_DENYLIST",
    "POLYMESH_CAPABILITY_PREFIX",
    "REVERSE_DNS_PREFIXES",
    "capability_name_from_skill_name",
    "is_publishable_skill",
    "map_capabilities_to_skills",
    "map_capability_to_skill",
    "map_card_from_a2a",
    "map_card_to_a2a",
    "skill_name_from_capability_name",
]
